unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java

```diff
@@ -938,11 +945,260 @@ public static void writeNormalizerTestSuite(String directory, String fileName)
+                int last = decomposition.codePointBefore(decomposition.length());
+                primaryCompositesByFirstNFDCodePoint
+                        .computeIfAbsent(first, key -> new TreeSet<>())
+                        .add(cp);
+                primaryCompositesByLastNFDCodePoint
+                        .computeIfAbsent(last, key -> new TreeSet<>())
+                        .add(cp);
+            }
+        }
+
+        Collection<String> skippedNFCs = new ArrayList<>();
+        Set<String> part5NFCs = new TreeSet<>();
+
+        // The set of all sequences of two code points appearing within a
+        // canonical decomposition.
+        Set<String> links = new TreeSet<>();
+        for (String decomposition : canonicalDecompositionsOfSingleCodepoints) {
+            int first = decomposition.codePointAt(0);
+            int second;
+            for (int i = UTF16.getCharCount(first);
+                    i < decomposition.length();
+                    first = second, i += UTF16.getCharCount(first)) {
+                second = decomposition.codePointAt(i);
+                links.add(Character.toString(first) + Character.toString(second));
+            }
+        }
+
+        for (String link : links) {
+            int first = link.codePointAt(0);
+            int second = link.codePointBefore(link.length());
+            // Look for primary composites firstCandidate and secondCandidate
+            // such that the concatenation of their canonical decompositions has
+            // the link at the boundary, e.g., for a link YZ, look for cases
+            // like firstDecomposition = XY, secondDecomposition = ZT.
+            // In addition to primary composites, allow the candidates to
+            // be a single canonically non-decomposable starters, thus
+            // firstDecomposition = XY, secondDecomposition = Z with ccc(Z)=0 or
+            // firstDecomposition = Y, secondDecomposition = ZT with ccc(Y)=0.
+            // We do not allow non-starters, since an initial nonstarter cannot
+            // do anything, and a final non-starter only exercises the CRA,
+            // which is not what we are looking for in Part 5.
+
+            // Note that if both the first and second candidates are
+            // canonically non-decomposable starters, any linking code point
+            // will cover the whole string, meaning this is already covered
+            // in Parts 1 and 4, so we do not emit any such test cases.
+            if (Default.ucd().getCombiningClass(first) == 0) {
+                primaryCompositesByLastNFDCodePoint
+                        .computeIfAbsent(first, key -> new TreeSet<>())
+                        .add(first);
+            }
+            if (Default.ucd().getCombiningClass(second) == 0) {
+                primaryCompositesByFirstNFDCodePoint
+                        .computeIfAbsent(second, key -> new TreeSet<>())
+                        .add(second);
+            }
```

Comment:
Should the ...ByFirst/ByLast maps be keyed by first vs. last code points? It's the opposite right now, and you are mapping each code point to itself.
If it is as intended, then I am confused... maybe a comment in that case?

Reply:
The actual building of the maps is above, ll. 975 _sqq_., where a comment indicates that I will cheesily add some more mappings of codepoints to themselves on top of what the identifier actually describes, which indeed happens here. I will try again to look for a cleaner way to write this.

Reply:
I have rewritten it without that cheesy mutability. PTAL.